Zebras: documentation and type fixes

abdulfatir opened the pull request:
*Issue #, if available:*

*Description of changes:*


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.


**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

Commit:
Add documentation

abdulfatir commented on src/zebras/_period.py:
```diff
@@ -339,7 +339,21 @@ def _encode_zebras_periods(v: Periods):
 
 
-def period(data, freq=None) -> Period:
+def period(data: Union[Period, str], freq: Optional[str] = None) -> Period:
```
@jaheba Still not sure what to do here. Should we import a pandas? Or modify the code to allow using zebras.Period as input too.

jaheba replied:
I think we should allow ``zebras.Period`` as well. Can we check if ``pandas.Period`` has some common base class that we could use here?

Otherwise, we just leave it undocumented for now, I think.

jaheba replied:
I think we should also allow to pass `zebras.Freq` as `freq`.

jaheba commented on src/zebras/_period.py:
```diff
@@ -339,7 +339,21 @@ def _encode_zebras_periods(v: Periods):
 
-def period(data, freq=None) -> Period:
+def period(data: Union[Period, str], freq: Optional[str] = None) -> Period:
+    """Creates a zebras.Period object that represents a period of time.
```
```suggestion
    """Create a zebras.Period object that represents a period of time.
```

Commit:
Allow using zebras.Period as input to Period

Commit:
Add docs to timeframe and timeseries

Commit:
Update period to accept Freq as freq

jaheba commented on src/zebras/_period.py:
```diff
@@ -64,6 +64,10 @@ class _BasePeriod:
 
+    @property
+    def freqstr(self) -> str:
+        return self.freq.name
```
```suggestion
        return str(self.freq)
```

`.name` does not include the number part, e.g. `zb.freq("3D").name == "D"`

jaheba commented on src/zebras/_period.py:
```diff
@@ -339,10 +343,29 @@ def _encode_zebras_periods(v: Periods):
+        else:
+            raise ValueError("No frequency specified.")
+    elif isinstance(freq, str):
         freq = Freq.from_pandas(freq)
```
I think we might want to add an explicit test for `Freq` here.

abdulfatir replied:
Updated but not sure about the `ValueError`. Is this okay?

jaheba replied:
I think `ValueError` is fine.

jaheba commented on src/zebras/_timeframe.py:
```diff
@@ -805,17 +805,56 @@ def with_index(self, index):
+    default_tdim: int = -1,
+    metadata: Optional[Dict] = None,
 ):
+    """Create a `zebras.TimeFrame` object that represents one
```
We should always use double backticks for sphinx when referencing types, etc:

```suggestion
    """Create a ``zebras.TimeFrame`` object that represents one
```

jaheba reviewed (commented):
Thanks for the improvements.

Commit:
Update freq stuff

Commit:
Add length to timeseries with post_init checks

Commit:
Add double backticks

Commit:
Fix to make tests work

abdulfatir commented:
@jaheba updated a few more things that need your review. Most importantly, added a `length` parameter to `TimeSeries` to be consistent with `TimeFrame` and added post init checks for the same. I will do split frame things in a separate PR.
